core/common/src/main/scala/org/typelevel/otel4s/Attributes.scala

```diff
@@ -112,25 +111,44 @@ sealed trait Attributes
 
   /** Removes all attributes with any of the given keys. */
   def removedAll(that: IterableOnce[AttributeKey[_]]): Attributes =
-    attributesFactory.fromSpecific((toMap -- that).values)
+    that.iterator.foldLeft(this)(_ - _)
 
   /** Removes all attributes with any of the given keys. */
   final def --(that: IterableOnce[AttributeKey[_]]): Attributes =
     removedAll(that)
 
-  /** @return the `Map` representation of these `Attributes` */
-  def toMap: Map[AttributeKey[_], Attribute[_]]
+  /** @return
+    *   the `Map` representation of these `Attributes` with untyped `String`
+    *   keys
+    */
+  def asMap: Map[String, Attribute[_]]
+
+  /** @return
+    *   the `Map` representation of these `Attributes` with untyped `String`
+    *   keys
+    */
+  @deprecated("use `asMap` instead", since = "0.5.0")
+  final def toMap: Map[String, Attribute[_]] = asMap
+
+  /** @return
+    *   the `Map` representation of these `Attributes` with typed
+    *   [[`AttributeKey`]]s
+    */
+  def asTypedMap: Map[AttributeKey[_], Attribute[_]] = new TypedMap
 
-  /** Equivalent to `toMap.keySet`.
+  /** Equivalent to `asTypedMap.keySet`.
     *
     * @return
     *   the keys of the [[`Attribute`]]s
     */
-  final def keys: Set[AttributeKey[_]] = toMap.keySet
+  final def keys: Set[AttributeKey[_]] = asTypedMap.keySet
```

Comment:
Should we have dual members: `keys` and `typedKeys`?

Reply:
I... don't think so, but the reason may not be straightforward.

fundamentally, `Attributes` is a `Map[String, Attribute[_]]` pretending to be a `Map[AttributeKey[_], Attribute[_]]`, and this inconsistency makes it difficult to have a coherent API. for example, `get` *must* take an `AttributeKey[_]` as a parameter and not just a `String` (for type safety), so to be consistent, `contains` ought to as well (as it should be equivalent to `get(_).isDefined`. I kind of wonder if we should have the `Map[String, Attribute[_]]` representation externally accessible at all.

there's also of course the alternative of going with `asUntypedMap` and `asMap`

Reply:
You are right, thanks for the explanation.